com.unity.ml-agents/Runtime/InferenceBrain/ApplierImpl.cs

```diff
@@ -13,18 +13,21 @@ namespace MLAgents.InferenceBrain
     /// </summary>
     public class ContinuousActionOutputApplier : TensorApplier.IApplier
     {
-        public void Apply(TensorProxy tensorProxy, IEnumerable<AgentIdActionPair> actions)
+        public void Apply(TensorProxy tensorProxy, IEnumerable<int> actionIds, Dictionary<int, float[]> lastActions)
         {
             var actionSize = tensorProxy.shape[tensorProxy.shape.Length - 1];
             var agentIndex = 0;
-            foreach (var idActionPair in actions)
+            foreach (int agentId in actionIds)
             {
                 var actionValue = new float[actionSize];
```

Comment:
We could avoid a lot of allocations by checking if `lastActions[agentId]` exists (and is the right size), and reusing it instead of `new`ing here.

Reply:
Unfortunately, no. The `lastActions` dictionary is not guaranteed to only contain the Agents requesting a decision at that particular step, nor is there a guarantee in the order of the keys.
Unless I misunderstood your comment?

Reply:
Making comments more inline.